<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to ZINB-WaVE • zinbwave</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to ZINB-WaVE">
<meta property="og:description" content="zinbwave">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">zinbwave</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.11.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro.html">An introduction to ZINB-WaVE</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/drisso/zinbwave/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="intro_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to ZINB-WaVE</h1>
                        <h4 class="author">Davide Risso</h4>
            
            <h4 class="date">Last modified: April 19, 2019; Compiled: July 10, 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drisso/zinbwave/blob/master/vignettes/intro.Rmd"><code>vignettes/intro.Rmd</code></a></small>
      <div class="hidden name"><code>intro.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{zinbwave Vignette}
-->
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>The recommended way to install the <code>zinbwave</code> package is</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw">BiocManager</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"zinbwave"</span>)
</pre></div>
<p>Note that <code>zinbwave</code> requires R (&gt;=3.4) and Bioconductor (&gt;=3.6).</p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This vignette provides an introductory example on how to work with the <code>zinbwave</code> package, which implements the ZINB-WaVE method proposed in <span class="citation">(Risso et al. 2018)</span>.</p>
<p>First, let’s load the packages and set serial computations.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">zinbwave</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">scRNAseq</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/HenrikBengtsson/matrixStats">matrixStats</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">biomaRt</span>)

<span class="co"># Register BiocParallel Serial Execution</span>
<span class="kw">BiocParallel</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span>(<span class="kw">BiocParallel</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></span>())
</pre></div>
<div id="the-zinb-wave-model" class="section level2">
<h2 class="hasAnchor">
<a href="#the-zinb-wave-model" class="anchor"></a>The ZINB-WaVE model</h2>
<p>ZINB-WaVE is a general and flexible model for the analysis of high-dimensional zero-inflated count data, such as those recorded in single-cell RNA-seq assays. Given <span class="math inline">\(n\)</span> samples (typically, <span class="math inline">\(n\)</span> single cells) and <span class="math inline">\(J\)</span> features (typically, <span class="math inline">\(J\)</span> genes) that can be counted for each sample, we denote with <span class="math inline">\(Y_{ij}\)</span> the count of feature <span class="math inline">\(j\)</span> (<span class="math inline">\(j=1,\ldots,J\)</span>) for sample <span class="math inline">\(i\)</span> (<span class="math inline">\(i=1,\ldots,n\)</span>). To account for various technical and biological effects, typical of single-cell sequencing technologies, we model <span class="math inline">\(Y_{ij}\)</span> as a random variable following a zero-inflated negative binomial (ZINB) distribution with parameters <span class="math inline">\(\mu_{ij}\)</span>, <span class="math inline">\(\theta_{ij}\)</span>, and <span class="math inline">\(\pi_{ij}\)</span>, and consider the following regression models for the parameters:</p>
<p><span class="math display">\[\begin{align}
\label{eq:model1}
\ln(\mu_{ij}) &amp;= \left( X\beta_\mu + (V\gamma_\mu)^\top + W\alpha_\mu + O_\mu\right)_{ij}\,,\\
\label{eq:model2}
\text{logit}(\pi_{ij}) &amp;= \left(X\beta_\pi + (V\gamma_\pi)^\top + W\alpha_\pi + O_\pi\right)_{ij} \,, \\
\label{eq:model3}
\ln(\theta_{ij}) &amp;= \zeta_j \,,
\end{align}\]</span>.</p>
<p>where the elements of the regression models are as follows.</p>
<ul>
<li>
<span class="math inline">\(X\)</span> is a known <span class="math inline">\(n \times M\)</span> matrix corresponding to <span class="math inline">\(M\)</span> cell-level covariates and <span class="math inline">\({\bf \beta}=(\beta_\mu,\beta_\pi)\)</span> its associated <span class="math inline">\(M \times J\)</span> matrices of regression parameters. <span class="math inline">\(X\)</span> can typically include covariates that induce variation of interest, such as cell types, or covariates that induce unwanted variation, such as batch or quality control (QC) measures. By default, it includes only a constant column of ones, <span class="math inline">\({\bf 1}_n\)</span>, to account for gene-specific intercepts.</li>
<li>
<span class="math inline">\(V\)</span> is a known <span class="math inline">\(J \times L\)</span> matrix corresponding to <span class="math inline">\(J\)</span> gene-level covariates, such as gene length or GC-content, and <span class="math inline">\({\bf \gamma} = (\gamma_\mu , \gamma_\pi)\)</span> its associated <span class="math inline">\(L\times n\)</span> matrices of regression parameters. By default, <span class="math inline">\(V\)</span> only includes a constant column of ones, <span class="math inline">\({\bf 1}_J\)</span>, to account for cell-specific intercepts, such as size factors representing differences in library sizes.</li>
<li>
<span class="math inline">\(W\)</span> is an unobserved <span class="math inline">\(n \times K\)</span> matrix corresponding to <span class="math inline">\(K\)</span> unknown cell-level covariates, which could be of “unwanted variation” or of interest (such as cell type), and <span class="math inline">\({\bf \alpha} = (\alpha_\mu,\alpha_{\pi})\)</span> its associated <span class="math inline">\(K \times J\)</span> matrices of regression parameters.</li>
<li>
<span class="math inline">\(O_\mu\)</span> and <span class="math inline">\(O_\pi\)</span> are known <span class="math inline">\(n \times J\)</span> matrices of offsets.</li>
<li>
<span class="math inline">\(\zeta\in\mathbb{R}^J\)</span> is a vector of gene-specific dispersion parameters on the log scale.</li>
</ul>
</div>
<div id="example-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#example-dataset" class="anchor"></a>Example dataset</h2>
<p>To illustrate the methodology, we will make use of the Fluidigm C1 dataset of <span class="citation">(Pollen et al. 2014)</span>. The data consist of 65 cells, each sequenced at high and low depth. The data are publicly available as part of the <a href="https://www.bioconductor.org/packages/release/data/experiment/html/scRNAseq.html">scRNAseq package</a>, in the form of a <code>SummarizedExperiment</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">fluidigm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ReprocessedData.html">ReprocessedFluidigmData</a></span>(assays = <span class="st">"tophat_counts"</span>)
<span class="kw">fluidigm</span>
</pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 26255 130 
## metadata(3): sample_info clusters which_qc
## assays(1): tophat_counts
## rownames(26255): A1BG A1BG-AS1 ... ZZEF1 ZZZ3
## rowData names(0):
## colnames(130): SRR1275356 SRR1274090 ... SRR1275366 SRR1275261
## colData names(28): NREADS NALIGNED ... Cluster1 Cluster2
## reducedDimNames(0):
## altExpNames(0):</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu">colData</span>(<span class="kw">fluidigm</span>)<span class="op">$</span><span class="kw">Coverage_Type</span>)
</pre></div>
<pre><code>## 
## High  Low 
##   65   65</code></pre>
</div>
</div>
<div id="gene-filtering" class="section level1">
<h1 class="hasAnchor">
<a href="#gene-filtering" class="anchor"></a>Gene filtering</h1>
<p>First, we filter out the lowly expressed genes, by removing those genes that do not have at least 5 reads in at least 5 samples.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="kw">filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="fu">assay</span>(<span class="kw">fluidigm</span>)<span class="op">&gt;</span><span class="fl">5</span>)<span class="op">&gt;</span><span class="fl">5</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="kw">filter</span>)
</pre></div>
<pre><code>## filter
## FALSE  TRUE 
## 16127 10128</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">fluidigm</span> <span class="op">&lt;-</span> <span class="kw">fluidigm</span>[<span class="kw">filter</span>,]
</pre></div>
<p>This leaves us with 10128 genes.</p>
<p>We next identify the 100 most variable genes, which will be the input of our ZINB-WaVE procedure. Although we apply ZINB-WaVE to only these genes primarily for computational reasons, it is generally a good idea to focus on a subset of highly-variable genes, in order to remove transcriptional noise and focus on the more biologically meaningful signals. However, at least 1,000 genes are probably needed for real analyses.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu">assay</span>(<span class="kw">fluidigm</span>) <span class="op">%&gt;%</span> <span class="kw">log1p</span> <span class="op">%&gt;%</span> <span class="kw">rowVars</span> <span class="op">-&gt;</span> <span class="kw">vars</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">vars</span>) <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">fluidigm</span>)
<span class="kw">vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw">vars</span>, decreasing = <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">vars</span>)
</pre></div>
<pre><code>##  IGFBPL1    STMN2     EGR1   ANP32E    CENPF     LDHA 
## 13.06109 12.24748 11.90608 11.67819 10.83797 10.72307</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="kw">fluidigm</span> <span class="op">&lt;-</span> <span class="kw">fluidigm</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">vars</span>)[<span class="fl">1</span><span class="op">:</span><span class="fl">100</span>],]
</pre></div>
<p>Before proceeding, we rename the first assay of <code>fluidigm</code> “counts” to avoid needing to specify which assay we should use for the <code>zinbwave</code> workflow. This is an optional step.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu">assayNames</span>(<span class="kw">fluidigm</span>)[<span class="fl">1</span>] <span class="op">&lt;-</span> <span class="st">"counts"</span>
</pre></div>
</div>
<div id="zinb-wave" class="section level1">
<h1 class="hasAnchor">
<a href="#zinb-wave" class="anchor"></a>ZINB-WaVE</h1>
<p>The easiest way to obtain the low-dimensional representation of the data with ZINB-WaVE is to use the <code>zinbwave</code> function. This function takes as input a <code>SummarizedExperiment</code> object and returns a <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">fluidigm_zinb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zinbwave.html">zinbwave</a></span>(<span class="kw">fluidigm</span>, K = <span class="fl">2</span>, epsilon=<span class="fl">1000</span>)
</pre></div>
<p>By default, the <code>zinbwave</code> function fits a ZINB model with <span class="math inline">\(X = {\bf 1}_n\)</span> and <span class="math inline">\(V = {\bf 1}_J\)</span>. In this case, the model is a factor model akin to principal component analysis (PCA), where <span class="math inline">\(W\)</span> is a factor matrix and <span class="math inline">\(\alpha_\mu\)</span> and <span class="math inline">\(\alpha_\pi\)</span> are loading matrices. By default, the <code>epsilon</code> parameter is set to the number of genes. We empirically found that a high <code>epsilon</code> is often required to obtained a good low-level representation. See <code><a href="../reference/zinbModel.html">?zinbModel</a></code> for details. Here we set <code>epsilon=1000</code>.</p>
<p>The parameter <span class="math inline">\(K\)</span> controls how many latent variables we want to infer from the data. <span class="math inline">\(W\)</span> is stored in the <code>reducedDim</code> slot of the object. (See the <code>SingleCellExperiment</code> vignette for details).</p>
<p>In this case, as we specified <span class="math inline">\(K=2\)</span>, we can visualize the resulting <span class="math inline">\(W\)</span> matrix in a simple plot, color-coded by cell-type.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">W</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span>(<span class="kw">fluidigm_zinb</span>)

<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">W</span>, bio=<span class="fu">colData</span>(<span class="kw">fluidigm</span>)<span class="op">$</span><span class="kw">Biological_Condition</span>,
           coverage=<span class="fu">colData</span>(<span class="kw">fluidigm</span>)<span class="op">$</span><span class="kw">Coverage_Type</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">W1</span>, <span class="kw">W2</span>, colour=<span class="kw">bio</span>, shape=<span class="kw">coverage</span>)) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(type = <span class="st">"qual"</span>, palette = <span class="st">"Set1"</span>) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>()
</pre></div>
<p><img src="intro_files/figure-html/zinb_plot-1.png" width="700"></p>
<div id="adding-covariates" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-covariates" class="anchor"></a>Adding covariates</h2>
<p>The ZINB-WaVE model is more general than PCA, allowing the inclusion of additional sample and gene-level covariates that might help to infer the unknown factors.</p>
<div id="sample-level-covariates" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-level-covariates" class="anchor"></a>Sample-level covariates</h3>
<p>Typically, one could include batch information as sample-level covariate, to account for batch effects. Here, we illustrate this capability by including the coverage (high or low) as a sample-level covariate.</p>
<p>The column <code>Coverage_Type</code> in the <code>colData</code> of <code>fluidigm</code> contains the coverage information. We can specify a design matrix that includes an intercept and an indicator variable for the coverage, by using the formula interface of <code>zinbFit</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="kw">fluidigm_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zinbwave.html">zinbwave</a></span>(<span class="kw">fluidigm</span>, K=<span class="fl">2</span>, X=<span class="st">"~Coverage_Type"</span>, epsilon=<span class="fl">1000</span>)
</pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="kw">W</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span>(<span class="kw">fluidigm_cov</span>)

<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">W</span>, bio=<span class="fu">colData</span>(<span class="kw">fluidigm</span>)<span class="op">$</span><span class="kw">Biological_Condition</span>,
           coverage=<span class="fu">colData</span>(<span class="kw">fluidigm</span>)<span class="op">$</span><span class="kw">Coverage_Type</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">W1</span>, <span class="kw">W2</span>, colour=<span class="kw">bio</span>, shape=<span class="kw">coverage</span>)) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(type = <span class="st">"qual"</span>, palette = <span class="st">"Set1"</span>) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>()
</pre></div>
<p><img src="intro_files/figure-html/zinb_plot2-1.png" width="700"></p>
<p>In this case, the inferred <span class="math inline">\(W\)</span> matrix is essentially the same with or without covariates, indicating that the scaling factor included in the model (the <span class="math inline">\(\gamma\)</span> parameters associated with the intercept of <span class="math inline">\(V\)</span>) are enough to achieve a good low-dimensional representation of the data.</p>
</div>
<div id="gene-level-covariates" class="section level3">
<h3 class="hasAnchor">
<a href="#gene-level-covariates" class="anchor"></a>Gene-level covariates</h3>
<p>Analogously, we can include gene-level covariates, as columns of <span class="math inline">\(V\)</span>. Here, we illustrate this capability by including gene length and GC-content.</p>
<p>We use the <code>biomaRt</code> package to compute gene length and GC-content.</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw">mart</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span>(<span class="st">"ensembl"</span>)
<span class="kw">mart</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useDataset.html">useDataset</a></span>(<span class="st">"hsapiens_gene_ensembl"</span>, mart = <span class="kw">mart</span>)
<span class="kw">bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span>(attributes=<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'hgnc_symbol'</span>, <span class="st">'start_position'</span>,
                         <span class="st">'end_position'</span>, <span class="st">'percentage_gene_gc_content'</span>),
            filters = <span class="st">'hgnc_symbol'</span>,
            values = <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">fluidigm</span>),
            mart = <span class="kw">mart</span>)

<span class="kw">bm</span><span class="op">$</span><span class="kw">length</span> <span class="op">&lt;-</span> <span class="kw">bm</span><span class="op">$</span><span class="kw">end_position</span> <span class="op">-</span> <span class="kw">bm</span><span class="op">$</span><span class="kw">start_position</span>
<span class="kw">len</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span>(<span class="kw">bm</span><span class="op">$</span><span class="kw">length</span>, <span class="kw">bm</span><span class="op">$</span><span class="kw">hgnc_symbol</span>, <span class="kw">mean</span>)
<span class="kw">len</span> <span class="op">&lt;-</span> <span class="kw">len</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">fluidigm</span>)]
<span class="kw">gcc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span>(<span class="kw">bm</span><span class="op">$</span><span class="kw">percentage_gene_gc_content</span>, <span class="kw">bm</span><span class="op">$</span><span class="kw">hgnc_symbol</span>, <span class="kw">mean</span>)
<span class="kw">gcc</span> <span class="op">&lt;-</span> <span class="kw">gcc</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">fluidigm</span>)]
</pre></div>
<p>We then include the gene-level information as <code>rowData</code> in the <code>fluidigm</code> object.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="fu">rowData</span>(<span class="kw">fluidigm</span>) <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(gccontent = <span class="kw">gcc</span>, length = <span class="kw">len</span>)
</pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="kw">fluidigm_gcc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zinbwave.html">zinbwave</a></span>(<span class="kw">fluidigm</span>, K=<span class="fl">2</span>, V=<span class="st">"~gccontent + log(length)"</span>, epsilon=<span class="fl">1000</span>)
</pre></div>
</div>
</div>
</div>
<div id="t-sne-representation" class="section level1">
<h1 class="hasAnchor">
<a href="#t-sne-representation" class="anchor"></a>t-SNE representation</h1>
<p>A t-SNE representation of the data can be obtained by computing the cell distances in the reduced space and running the t-SNE algorithm on the distance.</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">93024</span>)

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/jkrijthe/Rtsne">Rtsne</a></span>)
<span class="kw">W</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span>(<span class="kw">fluidigm_cov</span>)
<span class="kw">tsne_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html">Rtsne</a></span>(<span class="kw">W</span>, pca = <span class="fl">FALSE</span>, perplexity=<span class="fl">10</span>, max_iter=<span class="fl">5000</span>)

<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(Dim1=<span class="kw">tsne_data</span><span class="op">$</span><span class="kw">Y</span>[,<span class="fl">1</span>], Dim2=<span class="kw">tsne_data</span><span class="op">$</span><span class="kw">Y</span>[,<span class="fl">2</span>], 
           bio=<span class="fu">colData</span>(<span class="kw">fluidigm</span>)<span class="op">$</span><span class="kw">Biological_Condition</span>,
           coverage=<span class="fu">colData</span>(<span class="kw">fluidigm</span>)<span class="op">$</span><span class="kw">Coverage_Type</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">Dim1</span>, <span class="kw">Dim2</span>, colour=<span class="kw">bio</span>, shape=<span class="kw">coverage</span>)) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(type = <span class="st">"qual"</span>, palette = <span class="st">"Set1"</span>) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>()
</pre></div>
<p><img src="intro_files/figure-html/tsne-1.png" width="700"></p>
</div>
<div id="normalized-values-and-deviance-residuals" class="section level1">
<h1 class="hasAnchor">
<a href="#normalized-values-and-deviance-residuals" class="anchor"></a>Normalized values and deviance residuals</h1>
<p>Sometimes it is useful to have normalized values for visualization and residuals for model evaluation. Both quantities can be computed with the <code><a href="../reference/zinbwave.html">zinbwave()</a></code> function.</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="kw">fluidigm_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zinbwave.html">zinbwave</a></span>(<span class="kw">fluidigm</span>, K=<span class="fl">2</span>, epsilon=<span class="fl">1000</span>, normalizedValues=<span class="fl">TRUE</span>,
                    residuals = <span class="fl">TRUE</span>)
</pre></div>
<p>The <code>fluidigm_norm</code> object includes normalized values and residuals as additional <code>assays</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="kw">fluidigm_norm</span>
</pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 100 130 
## metadata(3): sample_info clusters which_qc
## assays(3): counts normalizedValues residuals
## rownames(100): IGFBPL1 STMN2 ... SRSF7 FAM117B
## rowData names(0):
## colnames(130): SRR1275356 SRR1274090 ... SRR1275366 SRR1275261
## colData names(28): NREADS NALIGNED ... Cluster1 Cluster2
## reducedDimNames(1): zinbwave
## altExpNames(0):</code></pre>
</div>
<div id="the-zinbfit-function" class="section level1">
<h1 class="hasAnchor">
<a href="#the-zinbfit-function" class="anchor"></a>The <code>zinbFit</code> function</h1>
<p>The <code>zinbwave</code> function is a user-friendly function to obtain the low-dimensional representation of the data, and optionally the normalized values and residuals from the model.</p>
<p>However, it is sometimes useful to store all the parameter estimates and the value of the likelihood. The <code>zinbFit</code> function allows the user to create an object of class <code>zinbModel</code> that can be used to store all the parameter estimates and have greater control on the results.</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="kw">zinb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zinbFit.html">zinbFit</a></span>(<span class="kw">fluidigm</span>, K=<span class="fl">2</span>, epsilon=<span class="fl">1000</span>)
</pre></div>
<p>As with <code>zinbwave</code>, by default, the <code>zinbFit</code> function fits a ZINB model with <span class="math inline">\(X = {\bf 1}_n\)</span> and <span class="math inline">\(V = {\bf 1}_J\)</span>.</p>
<p>If a user has run <code>zinbFit</code> and wants to obtain normalized values or the low-dimensional representation of the data in a <code>SingleCellExperiment</code> format, they can pass the <code>zinbModel</code> object to <code>zinbwave</code> to avoid repeating all the computations.</p>
<p>Here, we also specify <code>observationalWeights = TRUE</code> to compute observational weights, useful for differential expression (see next section).</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="kw">fluidigm_zinb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zinbwave.html">zinbwave</a></span>(<span class="kw">fluidigm</span>, fitted_model = <span class="kw">zinb</span>, K = <span class="fl">2</span>, epsilon=<span class="fl">1000</span>,
                          observationalWeights = <span class="fl">TRUE</span>)
</pre></div>
</div>
<div id="differential-expression" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential Expression</h1>
<p>The <code>zinbwave</code> package can be used to compute observational weights to “unlock” bulk RNA-seq tools for single-cell applications, as illustrated in <span class="citation">(Van den Berge et al. 2018)</span>.</p>
<p><code>zinbwave</code> optionally computes the observational weights when specifying <code>observationalWeights = TRUE</code> as in the code chuck above. See the man page of <code>zinbwave</code>. The weights are stored in an <code>assay</code> named <code>weights</code> and can be accessed with the following call.</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="kw">weights</span> <span class="op">&lt;-</span> <span class="fu">assay</span>(<span class="kw">fluidigm_zinb</span>, <span class="st">"weights"</span>)
</pre></div>
<p>Note that in this example, the value of the penalty parameter <code>epsilon</code> was set at <code>1000</code>, although we do not recommend this for differential expression analysis in real applications. Our evaluations have shown that a value of <code>epsilon=1e12</code> gives good performance across a range of datasets, although this number is still arbitrary. In general, values between <code>1e6</code> and <code>1e13</code> give best performances.</p>
<div id="differential-expression-with-edger" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-with-edger" class="anchor"></a>Differential expression with edgeR</h2>
<p>Once we have the observational weights, we can use them in <code>edgeR</code> to perform differential expression. Specifically, we use a moderated F-test in which the denominator residual degrees of freedom are adjusted by the extent of zero inflation (see <span class="citation">(Van den Berge et al. 2018)</span> for details).</p>
<p>Here, we compare NPC to GW16. Note that we start from only 100 genes for computational reasons, but in real analyses we would use all the expressed genes.</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span>)

<span class="kw">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span>(<span class="fu">assay</span>(<span class="kw">fluidigm_zinb</span>))
<span class="kw">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span>(<span class="kw">dge</span>)

<span class="kw">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(<span class="op">~</span><span class="kw">Biological_Condition</span>, data = <span class="fu">colData</span>(<span class="kw">fluidigm</span>))
<span class="kw">dge</span><span class="op">$</span><span class="kw">weights</span> <span class="op">&lt;-</span> <span class="kw">weights</span>
<span class="kw">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span>(<span class="kw">dge</span>, <span class="kw">design</span>)
<span class="kw">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html">glmFit</a></span>(<span class="kw">dge</span>, <span class="kw">design</span>)

<span class="kw">lrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/glmWeightedF.html">glmWeightedF</a></span>(<span class="kw">fit</span>, coef = <span class="fl">3</span>)
<span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span>(<span class="kw">lrt</span>)
</pre></div>
<pre><code>## Coefficient:  Biological_ConditionGW21+3 
##              logFC   logCPM       LR       PValue   padjFilter          FDR
## VIM      -4.768899 13.21736 47.44397 2.372950e-10 2.372950e-08 2.372950e-08
## FOS      -5.314329 14.50176 37.39308 8.937165e-09 4.468582e-07 4.468582e-07
## USP47    -3.900577 13.37158 29.91782 2.217492e-07 7.391640e-06 7.391640e-06
## PTN      -3.190168 13.22778 22.68392 4.691804e-06 1.172951e-04 1.172951e-04
## MIR100HG  2.388532 14.26683 18.25782 3.515256e-05 6.633079e-04 6.633079e-04
## NNAT     -2.062983 13.60868 17.90556 3.979848e-05 6.633079e-04 6.633079e-04
## SPARC    -3.202996 13.23879 16.08233 1.047878e-04 1.496968e-03 1.496968e-03
## SFRP1    -3.405951 13.01425 14.45825 2.439136e-04 3.048919e-03 3.048919e-03
## EGR1     -2.658644 14.93922 13.58180 3.193080e-04 3.547867e-03 3.547867e-03
## ST8SIA1  -3.338408 13.35883 12.64069 5.337675e-04 5.337675e-03 5.337675e-03</code></pre>
</div>
<div id="differential-expression-with-deseq2" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-with-deseq2" class="anchor"></a>Differential expression with DESeq2</h2>
<p>Analogously, we can use the weights in a <code>DESeq2</code> analysis by using observation-level weights in the parameter estimation steps. In this case, there is no need to pass the weights to <code>DESeq2</code> since they are already in the <code>weights</code> assay of the object.</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span>)

<span class="kw">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSet</a></span>(<span class="kw">fluidigm_zinb</span>, design = <span class="op">~</span> <span class="kw">Biological_Condition</span>)

<span class="kw">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span>(<span class="kw">dds</span>, sfType=<span class="st">"poscounts"</span>, useT=<span class="fl">TRUE</span>, minmu=<span class="fl">1e-6</span>)
<span class="kw">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html">lfcShrink</a></span>(<span class="kw">dds</span>, contrast=<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Biological_Condition"</span>, <span class="st">"NPC"</span>, <span class="st">"GW16"</span>),
                 type = <span class="st">"normal"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">res</span>)
</pre></div>
<pre><code>## log2 fold change (MAP): Biological_Condition NPC vs GW16 
## Wald test p-value: Biological Condition NPC vs GW16 
## DataFrame with 6 rows and 6 columns
##          baseMean log2FoldChange     lfcSE      stat      pvalue        padj
##         &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
## IGFBPL1  2054.403       -8.29483  0.705618  -8.14346 3.84143e-16 2.02181e-15
## STMN2    2220.078      -10.07429  0.769583  -9.37588 6.86037e-21 5.27721e-20
## EGR1     1342.465       -6.85394  0.662687 -10.34378 2.31566e-18 1.44729e-17
## ANP32E    806.983        1.99091  0.502374   3.96658 1.30782e-04 2.97232e-04
## CENPF     255.632        1.37109  0.553764   3.04964 3.19077e-03 5.80141e-03
## LDHA      311.760        2.36716  0.578059   4.91242 4.25408e-06 1.11170e-05</code></pre>
<p>Note that <code>DESeq2</code>’s default normalization procedure is based on geometric means of counts, which are zero for genes with at least one zero count. This greatly limits the number of genes that can be used for normalization in scRNA-seq applications. We therefore use the normalization method suggested in the <code>phyloseq</code> package, which calculates the geometric mean for a gene by only using its positive counts, so that genes with zero counts could still be used for normalization purposes. The <code>phyloseq</code> normalization procedure can be applied by setting the argument <code>type</code> equal to <code>poscounts</code> in <code>DESeq</code>.</p>
<p>For UMI data, for which the expected counts may be very low, the likelihood ratio test implemented in <code>nbinomLRT</code> should be used. For other protocols (i.e., non-UMI), the Wald test in <code>nbinomWaldTest</code> can be used, with null distribution a t-distribution with degrees of freedom corrected by the observational weights. In both cases, we recommend the minimum expected count to be set to a small value (e.g., <code>minmu=1e-6</code>).</p>
</div>
</div>
<div id="using-zinbwave-with-seurat" class="section level1">
<h1 class="hasAnchor">
<a href="#using-zinbwave-with-seurat" class="anchor"></a>Using <code>zinbwave</code> with Seurat</h1>
<p>The factors inferred in the <code>zinbwave</code> model can be added as one of the low dimensional data representations in the <code>Seurat</code> object, for instance to find subpopulations using Seurat’s cluster analysis method.</p>
<p>We first need to convert the <code>SingleCellExperiment</code> object into a <code>Seurat</code> object, using Seurat’s <code>CreateSeuratObject</code> function.</p>
<p>Note that the following workflow has been tested with Seurat’s version 3.0.0.</p>
<p>Here we create a simple Seurat object with the raw data. Please, refer to the Seurat’s vignettes for a typical analysis, which includes filtering, normalization, etc.</p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://www.satijalab.org/seurat">Seurat</a></span>)

<span class="kw">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/as.Seurat.html">as.Seurat</a></span>(x = <span class="kw">fluidigm_zinb</span>, counts = <span class="st">"counts"</span>, data = <span class="st">"counts"</span>)
</pre></div>
<p>Note that our <code>zinbwave</code> factors are automatically in the Seurat object.</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="kw">seu</span>
</pre></div>
<pre><code>## An object of class Seurat 
## 100 features across 130 samples within 1 assay 
## Active assay: RNA (100 features, 0 variable features)
##  1 dimensional reduction calculated: zinbwave</code></pre>
<p>Finally, we can use the <code>zinbwave</code> factors for cluster analysis.</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="kw">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="kw">seu</span>, reduction = <span class="st">"zinbwave"</span>,
                     dims = <span class="fl">1</span><span class="op">:</span><span class="fl">2</span> <span class="co">#this should match K</span>
                     )
<span class="kw">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(object = <span class="kw">seu</span>)
</pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 130
## Number of edges: 2461
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.7213
## Number of communities: 4
## Elapsed time: 0 seconds</code></pre>
</div>
<div id="working-with-large-datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#working-with-large-datasets" class="anchor"></a>Working with large datasets</h1>
<p>When working with large datasets, <code>zinbwave</code> can be computationally demanding. We provide an approximate strategy, implemented in the <code>zinbsurf</code> function, that uses only a random subset of the cells to infer the low dimensional space and subsequently projects all the cells into the inferred space.</p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="kw">fluidigm_surf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zinbsurf.html">zinbsurf</a></span>(<span class="kw">fluidigm</span>, K = <span class="fl">2</span>, epsilon = <span class="fl">1000</span>,
                          prop_fit = <span class="fl">0.5</span>)

<span class="kw">W2</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span>(<span class="kw">fluidigm_surf</span>)

<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">W2</span>, bio=<span class="fu">colData</span>(<span class="kw">fluidigm</span>)<span class="op">$</span><span class="kw">Biological_Condition</span>,
           coverage=<span class="fu">colData</span>(<span class="kw">fluidigm</span>)<span class="op">$</span><span class="kw">Coverage_Type</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">W1</span>, <span class="kw">W2</span>, colour=<span class="kw">bio</span>, shape=<span class="kw">coverage</span>)) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(type = <span class="st">"qual"</span>, palette = <span class="st">"Set1"</span>) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>()
</pre></div>
<p><img src="intro_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Note that here we use 50% of the data to get a reasonable approximation, since we start with only 130 cells. We found that for datasets with tens of thousands of cells, 10% (the default value) is usally a reasonable choice.</p>
<p>Note that this is an experimental feature and has not been thoroughly tested. Use at your own risk!</p>
</div>
<div id="a-note-on-performance-and-parallel-computing" class="section level1">
<h1 class="hasAnchor">
<a href="#a-note-on-performance-and-parallel-computing" class="anchor"></a>A note on performance and parallel computing</h1>
<p>The <code>zinbwave</code> package uses the <code>BiocParallel</code> package to allow for parallel computing. Here, we used the <code>register</code> command to ensure that the vignette runs with serial computations.</p>
<p>However, in real datasets, parallel computations can speed up the computations dramatically, in the presence of many genes and/or many cells.</p>
<p>There are two ways of allowing parallel computations in <code>zinbwave</code>. The first is to <code><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register()</a></code> a parallel back-end (see <code><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">?BiocParallel::register</a></code> for details). Alternatively, one can pass a <code>BPPARAM</code> object to <code>zinbwave</code> and <code>zinbFit</code>, e.g.</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span>)
<span class="kw">zinb_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/zinbwave.html">zinbwave</a></span>(<span class="kw">fluidigm</span>, K=<span class="fl">2</span>, BPPARAM=<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span>(<span class="fl">2</span>))
</pre></div>
<p>We found that <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam()</a></code> may have some performance issues on Mac; hence, we recommend <code><a href="https://rdrr.io/pkg/BiocParallel/man/DoparParam-class.html">DoparParam()</a></code> when working on Mac.</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
</pre></div>
<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Catalina 10.15.5
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] Seurat_3.1.5                DESeq2_1.29.6              
##  [3] edgeR_3.31.4                limma_3.45.7               
##  [5] Rtsne_0.15                  biomaRt_2.45.2             
##  [7] ggplot2_3.3.2               magrittr_1.5               
##  [9] scRNAseq_2.3.9              zinbwave_1.11.4            
## [11] SingleCellExperiment_1.11.6 SummarizedExperiment_1.19.5
## [13] DelayedArray_0.15.6         matrixStats_0.56.0         
## [15] Matrix_1.2-18               Biobase_2.49.0             
## [17] GenomicRanges_1.41.5        GenomeInfoDb_1.25.8        
## [19] IRanges_2.23.10             S4Vectors_0.27.12          
## [21] BiocGenerics_0.35.4         BiocStyle_2.17.0           
## 
## loaded via a namespace (and not attached):
##   [1] backports_1.1.8               AnnotationHub_2.21.1         
##   [3] BiocFileCache_1.13.0          plyr_1.8.6                   
##   [5] igraph_1.2.5                  lazyeval_0.2.2               
##   [7] splines_4.0.2                 BiocParallel_1.23.2          
##   [9] listenv_0.8.0                 digest_0.6.25                
##  [11] htmltools_0.5.0               memoise_1.1.0                
##  [13] cluster_2.1.0                 ROCR_1.0-11                  
##  [15] globals_0.12.5                annotate_1.67.0              
##  [17] askpass_1.1                   pkgdown_1.5.1.9000           
##  [19] prettyunits_1.1.1             colorspace_1.4-1             
##  [21] blob_1.2.1                    rappdirs_0.3.1               
##  [23] ggrepel_0.8.2                 xfun_0.15                    
##  [25] dplyr_1.0.0                   crayon_1.3.4                 
##  [27] RCurl_1.98-1.2                jsonlite_1.7.0               
##  [29] genefilter_1.71.0             survival_3.1-12              
##  [31] zoo_1.8-8                     ape_5.4                      
##  [33] glue_1.4.1                    gtable_0.3.0                 
##  [35] zlibbioc_1.35.0               XVector_0.29.3               
##  [37] leiden_0.3.3                  future.apply_1.6.0           
##  [39] scales_1.1.1                  DBI_1.1.0                    
##  [41] Rcpp_1.0.5                    viridisLite_0.3.0            
##  [43] xtable_1.8-4                  progress_1.2.2               
##  [45] reticulate_1.16               rsvd_1.0.3                   
##  [47] bit_1.1-15.2                  tsne_0.1-3                   
##  [49] htmlwidgets_1.5.1             httr_1.4.1                   
##  [51] RColorBrewer_1.1-2            ellipsis_0.3.1               
##  [53] ica_1.0-2                     pkgconfig_2.0.3              
##  [55] XML_3.99-0.4                  farver_2.0.3                 
##  [57] uwot_0.1.8                    dbplyr_1.4.4                 
##  [59] locfit_1.5-9.4                reshape2_1.4.4               
##  [61] tidyselect_1.1.0              labeling_0.3                 
##  [63] rlang_0.4.6                   softImpute_1.4               
##  [65] later_1.1.0.1                 AnnotationDbi_1.51.1         
##  [67] munsell_0.5.0                 BiocVersion_3.12.0           
##  [69] tools_4.0.2                   generics_0.0.2               
##  [71] RSQLite_2.2.0                 ExperimentHub_1.15.0         
##  [73] ggridges_0.5.2                evaluate_0.14                
##  [75] stringr_1.4.0                 fastmap_1.0.1                
##  [77] yaml_2.2.1                    knitr_1.29                   
##  [79] bit64_0.9-7                   fs_1.4.2                     
##  [81] fitdistrplus_1.1-1            purrr_0.3.4                  
##  [83] RANN_2.6.1                    pbapply_1.4-2                
##  [85] future_1.17.0                 nlme_3.1-148                 
##  [87] mime_0.9                      compiler_4.0.2               
##  [89] png_0.1-7                     plotly_4.9.2.1               
##  [91] curl_4.3                      interactiveDisplayBase_1.27.5
##  [93] tibble_3.0.2                  geneplotter_1.67.0           
##  [95] stringi_1.4.6                 desc_1.2.0                   
##  [97] lattice_0.20-41               vctrs_0.3.1                  
##  [99] pillar_1.4.4                  lifecycle_0.2.0              
## [101] BiocManager_1.30.10           lmtest_0.9-37                
## [103] RcppAnnoy_0.0.16              data.table_1.12.8            
## [105] cowplot_1.0.0                 bitops_1.0-6                 
## [107] irlba_2.3.3                   httpuv_1.5.4                 
## [109] patchwork_1.0.1               R6_2.4.1                     
## [111] bookdown_0.20                 promises_1.1.1               
## [113] gridExtra_2.3                 KernSmooth_2.23-17           
## [115] codetools_0.2-16              MASS_7.3-51.6                
## [117] assertthat_0.2.1              openssl_1.4.2                
## [119] rprojroot_1.3-2               withr_2.2.0                  
## [121] sctransform_0.2.1             GenomeInfoDbData_1.2.3       
## [123] hms_0.5.3                     grid_4.0.2                   
## [125] tidyr_1.1.0                   rmarkdown_2.3.2              
## [127] shiny_1.5.0</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Pollen2014">
<p>Pollen, Alex A, Tomasz J Nowakowski, Joe Shuga, Xiaohui Wang, Anne A Leyrat, Jan H Lui, Nianzhen Li, et al. 2014. “Low-coverage single-cell mRNA sequencing reveals cellular heterogeneity and activated signaling pathways in developing cerebral cortex.” <em>Nature Biotechnology</em> 32 (10): 1053–8.</p>
</div>
<div id="ref-risso2017">
<p>Risso, D, F Perraudeau, S Gribkova, S Dudoit, and Vert JP. 2018. “A General and Flexible Method for Signal Extraction from Single-Cell RNA-Seq Data.” <em>Nature Communications</em> 9: 284.</p>
</div>
<div id="ref-van2018observation">
<p>Van den Berge, Koen, Fanny Perraudeau, Charlotte Soneson, Michael I Love, Davide Risso, Jean-Philippe Vert, Mark D Robinson, Sandrine Dudoit, and Lieven Clement. 2018. “Observation Weights to Unlock Bulk Rna-Seq Tools for Zero Inflation and Single-Cell Applications.” <em>bioRxiv</em>, 250126.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Davide Risso, Svetlana Gribkova, Fanny Perraudeau, Jean-Philippe Vert, Clara Bagatin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
